### PR TITLE
Port tf2_kdl

### PR DESCRIPTION
**Pull request by vmayoral**

### CODE REVIEW

File: tf2_kdl/package.xml
Comment (tfoote) [inline tf2_kdl/package.xml:22]: Eigen is Eigen3 by default. 
https://index.ros.org/d/eigen/
If you want eigen2 you call it out explicitly
https://index.ros.org/d/eigen2/

But I think we've effectively transtioned to eigen3 fully now based on the rosindex lack of usage of the older key.

**Review by tfoote — COMMENTED** [review]
This would be great to have. Before merging this I'd like to see the unit tests ported too. They use a launch file in ROS1 but in ROS2 there's no master so that's not necessary for unit tests so everything can just be a unit test. The launch file only launches a single executable at a time anyway.

File: tf2_kdl/package.xml
Comment (vmayoral) [reply tf2_kdl/package.xml:22]: Understood. I recall having had some issues with that in OS X but can't reproduce it right now so I'll return to `eigen` in the package.xml.

**Comment by mlautman** [comment]
This port is still incomplete. Here are some issues that remain:

- [x] Finish package.xml port https://github.com/AcutronicRobotics/geometry2/pull/3
- [x] Port underlying [python_orcos_kdl](https://github.com/orocos/orocos_kinematics_dynamics/tree/master/python_orocos_kdl) for PyKDL 
- [ ] Remove roslib and rospy from [tf2_kdl](https://github.com/AcutronicRobotics/geometry2/blob/ros2/tf2_kdl/src/tf2_kdl/tf2_kdl.py)
- [ ] Remove deprecated [tf2_kdl/include/tf2_kdl/tf2_kdl subdirectory](https://github.com/AcutronicRobotics/geometry2/blob/ros2/tf2_kdl/include/tf2_kdl/tf2_kdl/tf2_kdl.h)
- [ ] Port the [CPP test](https://github.com/AcutronicRobotics/geometry2/blob/ros2/tf2_kdl/test/test_tf2_kdl.cpp) and python [test](https://github.com/AcutronicRobotics/geometry2/blob/ros2/tf2_kdl/scripts/test.py). Also, test.py should be moved to the test/ subdirectory instead of scripts and should not be built as an [executable](https://github.com/AcutronicRobotics/geometry2/blob/1020e530449ed2751d3cf7e52963ea9e6de1ee7b/tf2_kdl/CMakeLists.txt#L27) but as a python test.

File: tf2_kdl/CMakeLists.txt
Comment (mlautman) [inline tf2_kdl/CMakeLists.txt]: Why c++11? Why not 14. Add a note explaining this so that future devs know when it can be removed

File: tf2_kdl/CMakeLists.txt
Comment (mlautman) [inline tf2_kdl/CMakeLists.txt:15]: python libs need to be exported with  
```cmake 
ament_python_install_package(${PROJECT_NAME}
    PACKAGE_DIR src/${PROJECT_NAME})
```

File: tf2_kdl/CMakeLists.txt
Comment (mlautman) [inline tf2_kdl/CMakeLists.txt]: Don't just comment out tests. This port still needs work which gets hidden because the tests are disabled.

**Review by mlautman — CHANGES_REQUESTED** [review]

File: tf2_kdl/CMakeLists.txt
Comment (vmayoral) [reply tf2_kdl/CMakeLists.txt]: Thanks @mlautman 28d444e12eab3cc0c1be879426fbfa8f8cca48b2

**Comment by jacobperron** [comment]
@vmayoral I'm moved this back to "in progress", let me know if/when it's ready for another round of review.

**Comment by vmayoral** [comment]
Sure @jacobperron, will do.

**Comment by vmayoral** [comment]
@ahcorde cherry-picked your commits here. Please review.

**Comment by vmayoral** [comment]
@jacobperron, after @ahcorde's review, it should be ready to go from our side.

File: tf2_kdl/test/test_tf2_kdl.cpp
Comment (clalancette) [inline tf2_kdl/test/test_tf2_kdl.cpp]: I think this would be better as a `std::unique_ptr`.  Then we would make sure not to leak memory below.

File: tf2_kdl/test/test_tf2_kdl.cpp
Comment (clalancette) [inline tf2_kdl/test/test_tf2_kdl.cpp]: See comment below about making this a `std::unique_ptr`.

File: tf2_kdl/include/tf2_kdl/tf2_kdl.h
Comment (clalancette) [inline tf2_kdl/include/tf2_kdl/tf2_kdl.h]: We already have versions of this and `timeFromTimePoint` in https://github.com/ros2/geometry2/blob/ros2/tf2_ros/include/tf2_ros/buffer_interface.h#L46 ; is there any reason we can't use those?

**Review by clalancette — CHANGES_REQUESTED** [review]
There are a few things to fixup here, but nothing major.

File: tf2_kdl/include/tf2_kdl/tf2_kdl.h
Comment (ahcorde) [reply tf2_kdl/include/tf2_kdl/tf2_kdl.h]: Done, But I was wondering if this methods should go into `tf2/time.h`. inside `tf2_ros/buffer_interface.h` it's not very intuitive

Comment (clalancette) [reply tf2_kdl/include/tf2_kdl/tf2_kdl.h]: > Done, But I was wondering if this methods should go into `tf2/time.h`. inside `tf2_ros/buffer_interface.h` it's not very intuitive

The reason for that is that `tf2` itself is meant to be ROS-agnostic; you'll notice that there are no includes of ROS header files anywhere directly in tf2.  The interface between tf2 and ROS comes via the tf2_ros package.

That being said, I agree that it is not very intuitive that the time functions are in `tf2_ros/buffer_interface.h`.  I'd be in favor of making a new file `tf2_ros/time.h` where these methods would go, but I'll suggest we do that in a separate PR.

File: tf2_kdl/test/test_tf2_kdl.cpp
Comment (clalancette) [inline tf2_kdl/test/test_tf2_kdl.cpp]: With the change to a `unique_ptr`, this shouldn't be necessary; it will be destroyed as it goes out of scope.  So you can just remove this.

**Review by clalancette — CHANGES_REQUESTED** [review]
One more minor change, then this is good to go from me.  I'll run CI on it in the meantime.

**Comment by clalancette** [comment]
CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/6825/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/3121/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/5613/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/6651/)

**Comment by vmayoral** [comment]
@clalancette could you please consider https://github.com/ros2/geometry2/pull/93 as well (required for moveit2 as well)

File: tf2_eigen/include/tf2_eigen/tf2_eigen.h
Comment (clalancette) [inline tf2_eigen/include/tf2_eigen/tf2_eigen.h]: One other nit: need two spaces before the `return` here.

File: tf2_kdl/include/tf2_kdl/tf2_kdl.h
Comment (ahcorde) [reply tf2_kdl/include/tf2_kdl/tf2_kdl.h]: Sure! I like `tf2_ros/time.h`

**Comment by clalancette** [comment]
It looks like Windows is upset for two different reasons:

```
15:19:18 C:\J\workspace\ci_windows\ws\src\ros2\geometry2\tf2_kdl\test\test_tf2_kdl.cpp(45): error C2065: 'M_PI': undeclared identifier [C:\J\workspace\ci_windows\ws\build\tf2_kdl\test_kdl.vcxproj]
15:19:18 C:\J\workspace\ci_windows\ws\src\ros2\geometry2\tf2_kdl\test\test_tf2_kdl.cpp(120): warning C4244: '=': conversion from 'double' to 'builtin_interfaces::msg::Time_<std::allocator<void>>::_sec_type', possible loss of data [C:\J\workspace\ci_windows\ws\build\tf2_kdl\test_kdl.vcxproj]
15:19:18 C:\J\workspace\ci_windows\ws\src\ros2\geometry2\tf2_kdl\test\test_tf2_kdl.cpp(121): warning C4244: '=': conversion from 'double' to 'builtin_interfaces::msg::Time_<std::allocator<void>>::_nanosec_type', possible loss of data [C:\J\workspace\ci_windows\ws\build\tf2_kdl\test_kdl.vcxproj]
```

The M_PI thing we've dealt with before, but I can't exactly remember how; @tfoote , do you remember?  The loss of precision thing is actually true, so we either need to explicitly cast it or do something smarter there.

File: tf2_eigen/include/tf2_eigen/tf2_eigen.h
Comment (tfoote) [inline tf2_eigen/include/tf2_eigen/tf2_eigen.h]: Renormalizing the quaternion here will break the the one to one mapping I'd expect from a conversions. A -> B -> A should return the same thing.

File: tf2_kdl/test/test_tf2_kdl.cpp
Comment (tfoote) [inline tf2_kdl/test/test_tf2_kdl.cpp:53]: The windows M_PI warning can be worked around with an extra #define 

See https://github.com/ros2/geometry2/commit/dc7b000 for an example

File: tf2_eigen/include/tf2_eigen/tf2_eigen.h
Comment (tfoote) [inline tf2_eigen/include/tf2_eigen/tf2_eigen.h]: Same for renormalization above. The original form should be preserved.

**Review by tfoote — CHANGES_REQUESTED** [review]

File: tf2_kdl/test/test_tf2_kdl.cpp
Comment (clalancette) [reply tf2_kdl/test/test_tf2_kdl.cpp:53]: This should be resolved now (I'll run CI again to ensure it).

File: tf2_kdl/test/test_tf2_kdl.cpp
Comment (clalancette) [reply tf2_kdl/test/test_tf2_kdl.cpp]: This is still unresolved in the latest patches.

**Comment by clalancette** [comment]
Actually, before I run CI again, there hasn't been any change for the Windows loss of precision, so @ahcorde will need to do something about that before moving forward.

File: tf2_kdl/test/test_tf2_kdl.cpp
Comment (ahcorde) [reply tf2_kdl/test/test_tf2_kdl.cpp]: Done

**Comment by ahcorde** [comment]
can you try the CI with this last commit? 

`header.stamp` contains:

```
int32 sec
uint32 nanosec
```

I was assigning a double. Probably the sizeof(double) and the sizeof(int32) (or sizeof(uint32)) is different and that's why it's failing.

**Comment by clalancette** [comment]
Rebuild of just Windows: [Build Status](http://ci.ros2.org/job/ci_windows/6691/)

If this succeeds, I'll do one more CI run of everything.

**Comment by clalancette** [comment]
All right, Windows is happy, here is the whole lot:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/6871/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/3157/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/5644/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/6691/)

File: tf2_kdl/include/tf2_kdl/tf2_kdl.h
Comment (clalancette) [inline tf2_kdl/include/tf2_kdl/tf2_kdl.h]: We no longer need this now that we are using the other functions.

**Review by clalancette — COMMENTED** [review]
One last nitpick, then I'm happy to approve.

**Comment by clalancette** [comment]
All right, I removed the NSEC_PER_SECOND in 21a8551, and did a local build (the change is simple enough that I don't think CI is warranted here).  I'm going to approve, @tfoote I'll leave it to you to approve and merge.

**Review by clalancette — APPROVED** [review]

**Comment by tfoote** [comment]
Overall we're very close. The diff looks good. However I was just reading through the history and noticed that several items flagged by @mlautman as good things todo from earlier in this thread that have not been addressed: https://github.com/ros2/geometry2/pull/90#issuecomment-474060368

I checked off orocos_kdl as it's there. But the other ones should also be addressed, especially the rospy usage and deprecated header. The test porting would be great to do but that's improving beyond the existing functionality.

**Comment by ahcorde** [comment]
I have been trying to execute tf2_kdl tests in python. But right now there is no a official port in the main branch of `tf2_ros` and `tf2_py` or the action in `tf2_msgs`. I have been checking others PR like this: https://github.com/ros2/geometry2/pull/99. it seems that these libraries are working properly. 

But adding these libraries I'm still stuck on this line 

```
out = b.transform(tf2_ros.Stamped(v, builtin_interfaces.msg.Time(sec=2), 'a'), 'b')
```

because of this error:

```
Traceback (most recent call last):
  File "/home/ahcorde/moveit_ws/src/geometry2/tf2_kdl/scripts/test.py", line 29, in test_transform
    out = b.transform(tf2_ros.Stamped(v, builtin_interfaces.msg.Time(sec=2), 'a'), 'b')
  File "/home/ahcorde/moveit_ws/install/lib/python3.6/site-packages/tf2_ros/buffer_interface.py", line 64, in transform
    do_transform = self.registration.get(type(object_stamped))
  File "/home/ahcorde/moveit_ws/install/lib/python3.6/site-packages/tf2_ros/buffer_interface.py", line 203, in get
    raise TypeException('Type %s if not loaded or supported'% str(key))
tf2_ros.buffer_interface.TypeException: Type <class 'PyKDL.Vector'> if not loaded or supported
```

Any ideas?

**Comment by ahcorde** [comment]
I was able to fix the tf2_kdl test in python. But as I mentioned before I had to cherry pick some of the commits in this PR https://github.com/ros2/geometry2/pull/99.

Is there any chance to revisit this PR https://github.com/ros2/geometry2/pull/99 and merge it? otherwise this Python test is not going to work

**Comment by clalancette** [comment]
> Is there any chance to revisit this PR #99 and merge it? otherwise this Python test is not going to work

Sorry for being quiet here.  After a bit of discussion with @tfoote , I think we've determined that for now it will be fine to just not build and install the python stuff, and leave that as a TODO for later.  So we can forget about #99 in terms of this PR; I'll open an issue on this repository to finish that porting.

Towards getting things finished, I've taken the liberty of rebasing this PR and doing some cleanup in the CMakeLists.txt and package.xml to get things moving again.  Take a look at what I've done in ec63e8b, and let me know what you think.

I'm also going to run CI on this shortly.

**Comment by clalancette** [comment]
@vmayoral Apologies, we had a collision with 3bbfee556d5b8015d4c5f704005a9db5cb8b8d5e and my rebase.  I'll restore your changes in a few minutes.

**Comment by clalancette** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/7035/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/3270/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/5765/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/6854/)

**Comment by clalancette** [comment]
And issue for python port opened in https://github.com/ros2/geometry2/issues/110

**Comment by clalancette** [comment]
All right, CI is all green and things seem happy.  I'm going to give this 24 hours for any further objections, then I'll merge it.

**Comment by ahcorde** [comment]
@clalancette Thanks for reviewing it :+1:

**Comment by ros-discourse** [comment]
This pull request has been mentioned on **ROS Discourse**. There might be relevant details there:

https://discourse.ros.org/t/the-moveit-2-journey-part-1-porting-and-understanding-moveit-core/8718/1

**Pull request merged**
